loop.hpp

```diff
@@ -0,0 +1,489 @@
+                        break;
+                    func(it);
+                }
+                return it;
+            }
+        };
+    }
+
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename Begin, typename End, typename F>
+    HPX_FORCEINLINE Begin
+    loop(Begin begin, End end, F && f)
+    {
+        typedef typename std::iterator_traits<Begin>::iterator_category cat;
+        return detail::loop<cat>::call(begin, end, std::forward<F>(f));
+    }
+
+    template <typename Begin, typename End, typename CancelToken, typename F>
+    HPX_FORCEINLINE Begin
+    loop(Begin begin, End end, CancelToken& tok, F && f)
+    {
+        typedef typename std::iterator_traits<Begin>::iterator_category cat;
+        return detail::loop<cat>::call(begin, end, tok, std::forward<F>(f));
+    };
+
+
+    namespace detail
+    {
+        // Helper class to repeatedly call a function a given number of times
+        // starting from a given iterator position.
+
+        template <typename Iterator>
+        struct loop_n
+        {
+         /////////////////////////////////////////////////////////////////
+            // handle sequences of non-futures
+            template <typename Iter, typename F>
+            static Iter call(Iter it, std::size_t count, F && f)
+            {
+                for (/**/; count != 0; (void) --count, ++it)
+                    f(it);
+                return it;
+            }
+
+            template <typename Iter, typename CancelToken, typename F>
+            static Iter call(Iter it, std::size_t count, CancelToken& tok,
+                F && f)
+            {
+                for (/**/; count != 0; (void) --count, ++it)
+                {
+                    if (tok.was_cancelled())
+                        break;
+                    f(it);
+                }
+                return it;
+            }
+        };
+
+        template <typename Itr, typename ...TS>
+        struct loop_n <prefetching_iterator<Itr, TS...> >
+        {
+            /////////////////////////////////////////////////////////////////
+            // handle sequences of non-futures when using prefetching
+            template <typename F>
+            static prefetching_iterator<Itr, TS...>
+            call(prefetching_iterator<Itr, TS...> it,
+                std::size_t count, F && f)
+            {
+                Itr inner_it = it.base;
+                std::size_t cnt = count + 1;
+                for (/**/; cnt != 0; (void) --cnt, ++it)
+                {
+                    std::size_t j = it.idx;
+                    std::size_t last = it.idx + it.chunk_size;
+
+                    if (it.range_size < last)
+                        last = it.range_size;
+
+                    for(/**/; j < last ; ++j)
+                    {
+                        f(inner_it);
+                        ++inner_it;
+                    }
+
+#ifdef SUPPORT_mm_prefetch
```

Comment:
This is never set. It would be nice if this logic would be in prefetcher.hpp. Somthing like this:
```
#if defined(HPX_HAVE_X86_SSE_PREFETCH
template <typename Tuple>
void prefetch(Tuple &)
{
    int sequencer[] = {0, (/* do prefetch magic for x86*/, 0)...};
}
#elif defined(HPX_HAVE_SOME_OTHER_PREFETCH
template <typename Tuple>
void prefetch(Tuple &)
{
    int sequencer[] = {0, (/* do prefetch magic for other*/, 0)...};
}
#else
template <typename Tuple>
void prefetch(Tuple &)
{
    int sequencer[] = {0, (hpx::util::get<I>(t)[j+1], 0)...};
}
#endif
```